### PR TITLE
Enable user management domain model with events

### PR DESCRIPTION
**Pull request by sven1103**

### CODE REVIEW

File: domain/src/main/java/life/qbic/domain/usermanagement/UserDomainService.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+      throw new UserException("User with email address already exists.");
+    }
+    var domainEventProducer = DomainEventProducer.instance();
+    var user = User.create(fullName, email);
+    user.setPassword(rawPassword);
+    userRepository.addUser(user);
+
+    var userCreatedEvent = UserRegistered.createEvent(user.getId(), user.getFullName(),
+        user.getEmail());
+    domainEventProducer.publish(userCreatedEvent);
```
Comment (KochTobi) [inline domain/src/main/java/life/qbic/domain/usermanagement/UserDomainService.java:44]:
```suggestion
    var userRegisteredEvent = UserRegistered.create(user.getId(), user.getFullName(),
        user.getEmail());
    domainEventProducer.publish(userRegisteredEvent);
```

File: domain/src/main/java/life/qbic/domain/usermanagement/User.java [hunk 2]
```diff
@@ -178,11 +181,11 @@ public void setEmailConfirmed(boolean emailConfirmed) {
...
    * Checks if a given password is correct for a user
    *
-   * @param password Password that is being validated
+   * @param rawPassword Password that is being validated
    * @return true, if the given password is correct for the user
    */
-  public Boolean checkPassword(String password) {
-    return Objects.equals(PasswordEncryptionPolicy.create().encrypt(password), encryptedPassword);
+  public Boolean checkPassword(char[] rawPassword) {
+    return Objects.equals(PasswordEncryptionPolicy.create().encrypt(rawPassword), encryptedPassword);
```
Comment (KochTobi) [inline domain/src/main/java/life/qbic/domain/usermanagement/User.java:188]: Nice renaming! :+1:

File: domain/src/main/java/life/qbic/domain/events/DomainEventSubscriber.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package life.qbic.domain.events;
+
+/**
+ * <b><class short description - 1 Line!></b>
+ *
+ * <p><More detailed description - When to use, what it solves, etc.></p>
+ *
+ * @since <version tag>
+ */
```
Comment (KochTobi) [inline domain/src/main/java/life/qbic/domain/events/DomainEventSubscriber.java:9]: Documentation missing

File: domain/src/main/java/life/qbic/apps/datamanager/services/UserRegistrationService.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+ *
+ * @since <version tag>
+ */
+public class UserRegistrationService {
+
+  public UserRegistrationService() {
+    super();
+  }
+
+  public void registerNewUser(final String fullName, final String email, final char[] password) throws ServiceException {
```
Comment (KochTobi) [inline domain/src/main/java/life/qbic/apps/datamanager/services/UserRegistrationService.java:22]:
```suggestion
  public void registerUser(final String fullName, final String email, final char[] password) throws ServiceException {
```
I don't think you can register a user that already exists. So the method name can be shortened.

File: domain/src/main/java/life/qbic/apps/datamanager/services/UserRegistrationService.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    DomainEventProducer.instance().subscribe(new DomainEventSubscriber<UserRegistered>() {
+      @Override
+      public Class<UserRegistered> subscribedToEventType() {
+        return UserRegistered.class;
+      }
+
+      @Override
+      public void handleEvent(UserRegistered event) {
+        System.out.println("New user registered event: " + event.occurredOn().toString());
+        System.out.println(event.userId() + ": " + event.userEmail());
```
Comment (KochTobi) [inline domain/src/main/java/life/qbic/apps/datamanager/services/UserRegistrationService.java:33]:
```suggestion
        System.out.println("New user registered event: " + event.occurredOn().toString());
        System.out.println(event.userId() + ": " + event.userEmail());
        //TODO implement consequences of event occurence.
```

File: domain/src/main/java/life/qbic/domain/events/DomainEventProducer.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    }
+    try {
+      publishing.set(Boolean.TRUE);
+      List<DomainEventSubscriber<T>> registeredSubscribers = subscribers.get();
+      Class<? extends DomainEvent> domainEventType = domainEvent.getClass();
+      for (DomainEventSubscriber<T> subscriber : registeredSubscribers) {
+        if (subscriber.subscribedToEventType() == domainEventType) {
+          subscriber.handleEvent(domainEvent);
+        }
+      }
```
Comment (KochTobi) [inline domain/src/main/java/life/qbic/domain/events/DomainEventProducer.java:54]: NIT: personal flavor for readability
```suggestion
      registeredSubscribers.stream()
          .filter(it -> it.subscribedToEventType() == domainEventType)
          .forEach(DomainEventSubscriber::handleEvent):
```

File: domain/src/main/java/life/qbic/domain/usermanagement/DomainRegistry.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
+package life.qbic.domain.usermanagement;
+
+/**
+ * <b><class short description - 1 Line!></b>
+ *
+ * <p><More detailed description - When to use, what it solves, etc.></p>
+ *
+ * @since <version tag>
+ */
```
[the remark this answers is not in the document]
Comment (sven1103) [reply domain/src/main/java/life/qbic/domain/usermanagement/DomainRegistry.java:9]: This one resides in the domain only, and i dont want it to get polluted with third party framework libs.

File: domain/src/main/java/life/qbic/apps/datamanager/services/UserRegistrationService.java (hunk 4 above)
Comment (sven1103) [reply domain/src/main/java/life/qbic/apps/datamanager/services/UserRegistrationService.java:22]: Addressed now, thank you!

File: webapp/src/main/java/life/qbic/security/QBiCPasswordEncoder.java [hunk 8]
```diff
@@ -28,7 +28,7 @@ public QBiCPasswordEncoder() {
    */
   @Override
   public String encode(CharSequence rawPassword) {
-    return passwordEncryptionPolicy.encrypt(rawPassword.toString());
+    return passwordEncryptionPolicy.encrypt(rawPassword.toString().toCharArray());
```
[the remark this answers is not in the document]
Comment (sven1103) [reply webapp/src/main/java/life/qbic/security/QBiCPasswordEncoder.java:31]: The differenence between the char-array and a String is, that Strings are final. The content of the char array however is passed as reference within the application. This has the major advantage, that we only have one place in memory where the raw password is currently stored. With Strings, you create a lot of final copies. 

In order to retain control over the raw password, the client needs to clear the password once the registration has been executed. This is done by the UserRegistrationService currently (very rudimentary).

**Review by KochTobi — APPROVED** [review]
LGTM. Only one request (list typisation)

**Review by KochTobi — APPROVED** [review]
LGTM

**Pull request merged**